lib/hobbes/lang/preds/hasctor.C

```diff
@@ -5,33 +5,33 @@
 #include <hobbes/util/array.H>
 
 #include <hobbes/lang/preds/hasctor/variant.H>
+#include <memory>
 
 namespace hobbes {
 
-#define MSELECT_CTOR_FN "maybeFromCtor"
+static constexpr char const* MSELECT_CTOR_FN = "maybeFromCtor";
 
 /////////////////////////////////////////////////////
 // generic field verification and constraint removal
 /////////////////////////////////////////////////////
 CtorVerifier::CtorVerifier() {
-  this->eliminators.push_back(new HCVariantEliminator());
+  addEliminator(std::make_shared<HCVariantEliminator>());
 }
 
-std::string CtorVerifier::constraintName() {
-  return "HasCtor";
+const std::string& CtorVerifier::constraintName() {
+  static const auto n = std::string("HasCtor");
+  return n;
 }
 
-void CtorVerifier::addEliminator(HCEliminator* hce) {
+void CtorVerifier::addEliminator(const std::shared_ptr<HCEliminator>& hce) {
   this->eliminators.push_back(hce);
 }
 
 HCEliminator* CtorVerifier::findEliminator(const TEnvPtr& tenv, const HasCtor& hc, Definitions* ds) const {
-  for (auto hce : this->eliminators) {
-    if (hce->satisfiable(tenv, hc, ds)) {
-      return hce;
-    }
-  }
-  return 0;
+  const auto it = std::find_if(eliminators.begin(), eliminators.end(), [&](auto& hce) {
```

Comment:
what's the justification for this change?

Reply:
Just a minor refactoring, using standard algorithms is almost always easy to read and reason, IMO